api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/advancedhivillness/ListOfPatientsInAdvancedHivIllnessCohortQueries.java

```diff
@@ -0,0 +1,1403 @@
+            + " INNER JOIN encounter e ON e.patient_id = p.patient_id "
+            + "         INNER JOIN obs o ON o.encounter_id = e.encounter_id "
+            + "         INNER JOIN ( "
+            + listOfPatientsOnAdvancedHivIllnessQueries.getVLoadResultAndMostRecent()
+            + "         ) last_vl ON last_vl.patient_id = p.patient_id "
+            + " WHERE e.encounter_type = ${53} "
+            + " AND (o.concept_id = ${856} AND o.value_numeric IS NOT NULL) "
+            + " AND o.obs_datetime < last_vl.most_recent "
+            + " AND e.location_id = :location "
+            + " AND e.voided = 0 "
+            + " AND p.voided = 0 "
+            + " AND o.voided = 0 "
+            + " GROUP BY p.patient_id, o.value_numeric "
+            + " UNION "
+            + " SELECT p.patient_id, o.value_numeric AS viral_load, MAX(e.encounter_datetime) AS most_recent FROM patient p "
+            + " INNER JOIN encounter e ON e.patient_id = p.patient_id "
+            + "         INNER JOIN obs o ON o.encounter_id = e.encounter_id "
+            + "         INNER JOIN ( "
+            + listOfPatientsOnAdvancedHivIllnessQueries.getVLoadResultAndMostRecent()
+            + "         ) last_vl ON last_vl.patient_id = p.patient_id "
+            + " WHERE e.encounter_type IN(${6},${9},${13},${51}) "
+            + " AND (o.concept_id = ${856} AND o.value_numeric IS NOT NULL) "
+            + " AND e.encounter_datetime < last_vl.most_recent "
+            + " AND e.location_id = :location "
+            + " AND e.voided = 0 "
+            + " AND p.voided = 0 "
+            + " AND o.voided = 0 "
+            + " GROUP BY p.patient_id, o.value_numeric "
+            + " ) AS result_date GROUP BY result_date.patient_id ";
+
+    StringSubstitutor stringSubstitutor = new StringSubstitutor(map);
+
+    sqlPatientDataDefinition.setQuery(stringSubstitutor.replace(query));
+
+    return sqlPatientDataDefinition;
+  }
+
+  /**
+   * <b>Resultado do Última TB-LAM / Data do Último TB-LAM </b>
+   * <li>O registo mais recente de resultado TB-LAM= “positivo, negativo, NA, NF”, na “Ficha Clínica
+   *     –Ficha Mestra” ou “Ficha de Laboratório” ou “Ficha de Doença Avançada por HIV” ocorrido até
+   *     o fim do período de avaliação. or
+   *
+   *     <p>A data do registo mais recente de resultado de TB-LAM , na “Ficha Clínica –Ficha Mestra”
+   *     ou “Ficha de Laboratório” ou “Ficha de Doença Avançada por HIV” ocorrido até o fim do
+   *     período de avaliação.
+   *
+   * @param encounterTypeList EncounterTypes to be evaluated
+   * @param examConceptList Exam Concepts
+   * @param resultConceptList Result Concepts
+   * @param examResult Flag to return Exam result or Date
+   * @return {@link DataDefinition}
+   */
+  public DataDefinition getTbLaboratoryResearchResults(
+      List<Integer> encounterTypeList,
+      List<Integer> examConceptList,
+      List<Integer> resultConceptList,
+      boolean examResult) {
+
```

Comment:
Why not list of encounters and concept instead of Integers

Reply:
All changes indicated on the comments were applied!